File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
  def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
    candidates.sort()
    ans = []

    def backtrack(j, target, path):
      if target < 0:
        return
      if target == 0:
        ans.append(path)
        return
      for i in range(j, len(candidates)):
        if i > j and candidates[i] == candidates[i - 1]:
          continue
        backtrack(i + 1, target - candidates[i], path + [candidates[i]])

    backtrack(0, target, [])

    return ans

File: 0040-combination-sum-ii/0040-combination-sum-ii.go
func combinationSum2(candidates []int, target int) [][]int {
	slices.Sort(candidates)
	ans := [][]int{}
	var backtrack func(i int, target int, path []int)
	backtrack = func(i int, target int, path []int) {
		if target < 0 {
			return
		} else if target == 0 {
			ans = append(ans, slices.Clone(path))
			return
		}
		for j := i; j < len(candidates); j++ {
			if j > i && candidates[j] == candidates[j-1] {
				continue
			}
      path = append(path, candidates[j])
			backtrack(j+1, target-candidates[j], path)
      path = path[:len(path)-1]
		}
	}

	backtrack(0, target, []int{})
	return ans
}